alf/pretrained_models/model_adapters.py

```diff
@@ -261,18 +268,51 @@ def can_adapt(cls, m: nn.Module) -> bool:
     @classmethod
     def _decompose_weight_dims(cls, m: nn.Module):
         kernel_size = m.kernel_size
-        if not isinstance(kernel_size, tuple):
-            kernel_size = (kernel_size, ) * 2
-        nrows = m.out_channels // m.groups * kernel_size[0]
-        ncols = m.in_channels * kernel_size[1]
+        # conv2d weight has a shape: ``(out_channels, in_channels // groups, *kernel_size)``
+        nrows = m.out_channels * kernel_size[0]
+        ncols = m.in_channels // m.groups * kernel_size[1]
         return nrows, ncols
 
     def forward(self, input):
         m = self._m[0]
-        return F.conv2d(
-            input,
-            self._adapter_weight(),
-            stride=m.stride,
-            padding=m.padding,
-            dilation=m.dilation,
-            groups=m.groups)
+        if isinstance(m.padding, str) or m.groups > 1 or self._wB is None:
+            # Three senarios when two-stage convolution is difficult:
+            # 1. m.groups > 1: r has to be divisible by m.groups in order to
+            #    preserve the correct input-output mapping
+            # 2. m.padding is a string: torch will compute paddings on the fly,
+            #    so we won't know its values in advance.
+            # 3. low-rank decomposition is not available
+            return F.conv2d(
+                input,
+                self._adapter_weight(),
+                stride=m.stride,
+                padding=m.padding,
+                dilation=m.dilation,
+                groups=m.groups)
+        else:
+            input = F.conv2d(
+                input,
+                self._wA.reshape(self._r, m.in_channels, 1, m.kernel_size[1]),
+                stride=(1, m.stride[1]),
```

Comment:
Not sure if this results in fewer computations as we have stride=1 for H in the first convolution?